hyperspy/_signals/sed.py

```diff
@@ -0,0 +1,465 @@
+            sed_par.scan_rotation,
+            'Acquisition_instrument.TEM.convergence_angle':
+            sed_par.convergence_angle,
+            'Acquisition_instrument.TEM.precession_angle':
+            sed_par.precession_angle,
+            'Acquisition_instrument.TEM.precession_frequency':
+            sed_par.precession_frequency,
+            'Acquisition_instrument.TEM.Detector.SED.exposure_time':
+            sed_par.exposure_time, }
+
+        for key, value in mapping.iteritems():
+            if value != t.Undefined:
+                self.metadata.set_item(key, value)
+        self._are_microscope_parameters_missing()
+
+    def _are_microscope_parameters_missing(self):
+        """Check that the SED parameters necessary for pattern calibration are
+        defined in metadata and raise a UI if not to add them."""
+        must_exist = ('Acquisition_instrument.TEM.beam_energy',
+                      'Acquisition_instrument.TEM.camera_length',
+                      'Acquisition_instrument.TEM.scan_rotation')
+
+        missing_parameters = []
+        for item in must_exist:
+            exists = self.metadata.has_item(item)
+            if exists is False:
+                missing_parameters.append(item)
+        if missing_parameters:
+            if preferences.General.interactive is True:
+                par_str = "The following parameters are missing:\n"
+                for par in missing_parameters:
+                    par_str += '%s\n' % par
+                par_str += 'Please set them in the following wizard'
+                is_ok = messagesui.information(par_str)
+                if is_ok:
+                    self._set_microscope_parameters()
+                else:
+                    return True
+            else:
+                return True
+        else:
+            return False
+
+    def get_direct_beam_mask(self, radius=None, center=None):
+        """Generate a signal mask for the direct beam.
+
+        Parameters
+        ----------
+        radius : int
+            User specified radius for the circular mask.
+        center : tuple, None
+            User specified (x, y) position of the diffraction pattern center.
+            i.e. the direct beam position. If None it is assumed that the direct
+            beam is at the center of the diffraction pattern.
+
+        Return
+        ------
+        mask : array
+            The mask of the direct beam
+        """
+        r = radius
+        ny = self.axes_manager.signal_shape[1]
+        nx = self.axes_manager.signal_shape[0]
+
+        if center == None:
+            a = (self.axes_manager.signal_shape[1] - 1) / 2
+            b = (self.axes_manager.signal_shape[0] - 1) / 2
+        else:
+            a, b = center
+
+        y, x = np.ogrid[-b:ny-b, -a:nx-a]
+        mask = x*x + y*y <= r*r
+        return mask
+
+    def get_direct_beam_position(self, radius=None):
+        """Estimate the position of the direct beam in each SED pattern.
+
```

Comment:
I don't think that they do? The idea is that because the direct beam shouldn't change position too much in a dataset summing reinforces this more than all other diffracted beams so in the sum we're relatively safe to guess that the maximum intensity pixel is the direct beam. However, it might move a bit so it then goes through and looks for more intense pixels near to the initial guess and if it finds one moves the center to there. I then added in the possibility to find many equal maxima to ensure that things still work IF its saturated - but if it's not, it should still work well -- in fact the tests make non-saturated models to illustrate this.